src/Interpreters/Cache/FileCache.cpp

```diff
@@ -1727,6 +1732,57 @@ void FileCache::applySettingsIfPossible(const FileCacheSettings & new_settings,
 
                     /// Do actual eviction from filesystem.
                     eviction_candidates.evict();
+
+                    auto failed_candidates = eviction_candidates.getFailedCandidates();
+                    if (failed_candidates.total_cache_size)
+                    {
+                        actual_settings[FileCacheSetting::max_size] = std::min<size_t>(
+                            actual_settings[FileCacheSetting::max_size].value,
+                            new_settings[FileCacheSetting::max_size] + failed_candidates.total_cache_size);
+
+                        actual_settings[FileCacheSetting::max_elements] = std::min<size_t>(
+                            actual_settings[FileCacheSetting::max_elements].value,
+                            new_settings[FileCacheSetting::max_elements] + failed_candidates.total_cache_elements);
+
+                        cache_lock.lock();
+
+                        /// Increase the max size and max elements
+                        /// to the size and number of failed candidates.
+                        main_priority->modifySizeLimits(
+                            actual_settings[FileCacheSetting::max_size],
+                            actual_settings[FileCacheSetting::max_elements],
+                            new_settings[FileCacheSetting::slru_size_ratio],
+                            cache_lock);
+
+                        /// Add failed candidates back to queue.
+                        for (const auto & [key_metadata, key_candidates] : failed_candidates.failed_candidates_per_key)
+                        {
+                            auto locked_key = key_metadata->tryLock();
```

Comment:
if `key_candidates` is empty we can skip this completely

Reply:
I will add `chassert(!key_candidates.empty())` and make sure it cannot be empty (now it can, I missed this part, thanks for pointing it out)